pkg/publicshare/manager/cs3/cs3.go

```diff
@@ -0,0 +1,401 @@
+	}
+
+	s := &PublicShareWithPassword{
+		PublicShare: &link.PublicShare{
+			Id:                id,
+			Owner:             ri.GetOwner(),
+			Creator:           u.Id,
+			ResourceId:        ri.Id,
+			Token:             tkn,
+			Permissions:       g.Permissions,
+			Ctime:             createdAt,
+			Mtime:             createdAt,
+			PasswordProtected: passwordProtected,
+			Expiration:        g.Expiration,
+			DisplayName:       displayName,
+		},
+		HashedPassword: password,
+	}
+
+	err := m.persist(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.PublicShare, nil
+}
+
+// UpdatePublicShare updates an existing public share
+func (m *Manager) UpdatePublicShare(ctx context.Context, u *user.User, req *link.UpdatePublicShareRequest) (*link.PublicShare, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	ps, err := m.getWithPassword(ctx, req.Ref)
+	if err != nil {
+		return nil, err
+	}
+
+	switch req.Update.Type {
+	case link.UpdatePublicShareRequest_Update_TYPE_DISPLAYNAME:
+		ps.PublicShare.DisplayName = req.Update.DisplayName
+	case link.UpdatePublicShareRequest_Update_TYPE_PERMISSIONS:
+		ps.PublicShare.Permissions = req.Update.Grant.Permissions
+	case link.UpdatePublicShareRequest_Update_TYPE_EXPIRATION:
+		ps.PublicShare.Expiration = req.Update.Grant.Expiration
+	case link.UpdatePublicShareRequest_Update_TYPE_PASSWORD:
+		h, err := bcrypt.GenerateFromPassword([]byte(req.Update.Grant.Password), m.passwordHashCost)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not hash share password")
+		}
+		ps.HashedPassword = string(h)
+		ps.PublicShare.PasswordProtected = true
+	default:
+		return nil, errtypes.BadRequest("no valid update type given")
+	}
+
+	err = m.persist(ctx, ps)
+	if err != nil {
+		return nil, err
+	}
+
+	return ps.PublicShare, nil
+}
+
+// GetPublicShare returns an existing public share
+func (m *Manager) GetPublicShare(ctx context.Context, u *user.User, ref *link.PublicShareReference, sign bool) (*link.PublicShare, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	ps, err := m.getWithPassword(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+
+	if ps.PublicShare.PasswordProtected && sign {
+		err = publicshare.AddSignature(ps.PublicShare, ps.HashedPassword)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ps.PublicShare, nil
+}
+
+func (m *Manager) getWithPassword(ctx context.Context, ref *link.PublicShareReference) (*PublicShareWithPassword, error) {
+	switch {
+	case ref.GetToken() != "":
+		return m.getByToken(ctx, ref.GetToken())
+	case ref.GetId().GetOpaqueId() != "":
+		return m.getByID(ctx, ref.GetId().GetOpaqueId())
+	default:
+		return nil, errtypes.BadRequest("neither id nor token given")
+	}
+}
+
+func (m *Manager) getByID(ctx context.Context, id string) (*PublicShareWithPassword, error) {
+	tokens, err := m.indexer.FindBy(&link.PublicShare{}, "Id.OpaqueId", id)
+	if err != nil {
+		return nil, err
+	}
+	if len(tokens) == 0 {
```

Comment:
Should we consider the case `len(tokens) > 1`? I could imagine debugging will be hard in that case.

Reply:
Hm, the index being used is a "unique" index so it'll never return more than one tokens. And *if* there were more than one they would (hopefully!) be identical in which case it would be fine to just pick one. I would vote for not bothering in this case :)

Reply:
Ok. Fine for me :+1: